Add Funcbench feature and update the readme files. (#257)
prometheus/test-infra: funcbench/README.md

## Intro

> A tool used as a github action to run a `go test -bench` and compare changes from a PR against another branch. The benchmark is triggered by creating a comment which specifies a branch to compare. The results are then posted back into the PR comments. Comparison is done using [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp). Arguments for the benchcmp are read from files created by previous action (for example [commentMonitor](/tools/commentMonitor/main.go)), which is responsible for the comment parsing.

Reviewer: ```suggestion
A tool used as a github action to run a `go test -bench` and compare changes from a PR against another branch. The benchmark is triggered by creating a comment which specifies a branch to compare. The results are then posted back as a PR comment. 
Comparison is done using [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp). Arguments for the benchcmp are read from files created by previous action (for example [commentMonitor](/tools/commentMonitor/main.go)), which is responsible for the comment parsing.
```

## Use

> After a successful set up, you can create a comment in a pull request to execute the benchmarks. The form of the comment is following:

Reviewer: ```suggestion
Tests are triggered by posting a comment in a PR with the following format:
```

`/funcbench [branch] <golang test regex>  [-no-race]`

> If no branch is specified, `master` branch is used. The golang test regex is used to execute benchmarks that match the regex. By defualt the benchmarks run with `-race` flag enabled, disable this by specifying `-no-race` in the comment. After a successful run, a table with results will be displayed in the comment section.

Reviewer: ```suggestion
Specifying which tests to run are filtered by using the standard golang regex format. 
By default all benchmarks run with `-race` flag enabled and it can be disabled by appending `-no-race` at the end of the comment. 
After a successful run a github comment is created with the results.
```

> ## Intro

Reviewer: https://github.com/prometheus-community/tsdb/pull/1/files 

What about the version I suggested?

also can you add an example how to build the docker image.

Author: I added every new idea from your suggestion.

"How to build image" seciton is 16 lines under this comment.

Reviewer: In a readme I would expect to know what the tool does by just reading the first 1-2 sentences. In the current readme it is not quite clear until the second paragraph.

In the format I suggested it is quite clear what is the workflow and makes it more clear that it expects some arguments that are passed by another action.

This tools is meant to be used as a Github action. The action itself is, to a large degree, unusable alone, as you need to combine it with another Github action that will provide necessary files to it. At this time, the only action it is supposed to work with, is [comment-monitor](https://github.com/prometheus/prombench/tree/master/tools/commentMonitor).

## Use
After a successful set up, you can create a comment in a pull request to execute the benchmarks. The form of the comment is following:
`/benchmark [branch] <golang test regex>  [-no-race]`
If no branch is specified, `master` branch is used. The golang test regex is used to execute benchmarks that match the regex. By defualt the benchmarks run with `-race` flag enabled, disable this by specifying `-no-race` in the comment. After a successful run, a table with results will be displayed in the comment section.

## Set up
- Create Github actions workflow file that is executed when an issue comment is created, `on = "issue_comment"`.
- Add comment-monitor Github action as a first step.

> - Specify this regex `^/benchmark ?([^ B\.]+)? ?(\.|Bench.*|[^ ]+)? ?(-no-race)?.*$` in the `args` field of the comment-monitor.

Reviewer: ```suggestion
- Specify this regex `^/funcbench ?([^ B\.]+)? ?(\.|Bench.*|[^ ]+)? ?(-no-race)?.*$` in the `args` field of the comment-monitor.
```

> - Specify this Github action as a pre-built image, build from this source code, or just refer to this repository from the workflow file.

Reviewer: It would be much more clear if you just show the example first with comments for what each env var is for.

- Provide a Github token as an environment variable to both comment-monitor and funcbench.

## How to build
`docker build -t <tag of your choice> .`

## Examples

> ### Example comments

Reviewer: An example how to run a bench test by an exact func name would also be useful.

Author: I hoped it would evident from the fact that we use regex to match the functions.

Reviewer: 3 examples:
Bench a func by an exact match
Bench all functions with the same suffix
Bench all functions from all modules.

Execute all benchmarks matching `FuncName` regex, and compare it with `master` branch.
 - `/benchmark master FuncName.*`

Execute all benchmarks, and compere the results with `devel` branch.
 - `/benchmark devel .`

### Example Github actions workflow file
```
on: issue_comment
name: Benchmark
jobs:
  commentMonitor:
    runs-on: ubuntu-latest
    steps:
    - name: commentMonitor
      uses: docker://prombench/comment-monitor:latest
      env:
        COMMENT_TEMPLATE: 'The benchmark has started.'
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        args: '"^/benchmark ?([^ ]+)? ?([^ ]+)? ?(-no-race)?.*$"'
    - name: benchmark
      uses: docker://prombench/funcbench:latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

> ## What it does [SHORT]

Reviewer: As I mentioned lets put the overview at the top.

It acquires arguments for `go test` benchmarks, executes and compares the benchmarks based on the provided arguments. Then it posts results into pull request it was called from.

> ## What it does [LONG]

Reviewer: It is only an additional sentence, why not combine it in a single paragraph?

Author: I wanted to add more, but we switched to another task and I forgot what I wanted to say.

## Intro
A tool used as a github action to run a `go test -bench` and compare changes from a PR against another branch. The benchmark is triggered by creating a comment which specifies a branch to compare. The results are then posted back as a PR comment. 
Comparison is done using [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp). Arguments for the benchcmp are read from files created by previous action (for example [commentMonitor](/tools/commentMonitor/main.go)), which is responsible for the comment parsing.

## Use
Tests are triggered by posting a comment in a PR with the following format:

> `/funcbench <branch> <golang test regex>  [-no-race]`

Reviewer: use ``` and a new line to make it on a new line

Specifying which tests to run are filtered by using the standard golang regex format. 
By default all benchmarks run with `-race` flag enabled and it can be disabled by appending `-no-race` at the end of the comment. 

> After a successful run a github comment is created with the results.

Reviewer: ```suggestion
```

this is already mentioned in the first paragraph

> ### Example Github actions workflow file

Reviewer: not sure if we should keep this example file or just put a link to the one that will be in the prometheus repo.
Having it here means that when we change things need to also update the example where if we just put a link to the prometheus repo it will always be up to date.

Author: The example, is just an example. As long as it is a valid example I think it is okay, and it does not have to be directing to something useful. I think we should modify the example only if we modify the `funcbench`. Also, it might be possible that the link will be broken and we might never find out. So I would rather keep it as in the README as a text.

Reviewer: ok

    - name: benchmark
      uses: docker://prombench/funcbench:latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} // Github secret token
```

## Set up
This tools is meant to be used as a Github action. The action itself is, to a large degree, unusable alone, as you need to combine it with another Github action that will provide necessary files to it. At this time, the only action it is supposed to work with, is [comment-monitor](https://github.com/prometheus/prombench/tree/master/tools/commentMonitor).
- Create Github actions workflow file that is executed when an issue comment is created, `on = "issue_comment"`.
- Add comment-monitor Github action as a first step.
- Specify this regex `^/funcbench ?(?P<BRANCH>[^ B\.]+)? ?(?P<REGEX>\.|Bench.*|[^ ]+)? ?(?P<RACE>-no-race)?.*$` in the `args` field of the comment-monitor.
- Specify this Github action as a pre-built image, build from this source code, or just refer to this repository from the workflow file.
- Provide a Github token as an environment variable to both comment-monitor and funcbench.

## How to build
`docker build -t <tag of your choice> .`

## Examples

> ### Example comments

Reviewer: ```suggestion
```
